core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -679,6 +689,11 @@ else if (!found && result) {
                     return new ComparisonExpression(ComparisonExpression.Operator.EQUAL, toExpression(value, type), simplifiedExpressionValues.get(0));
                 }
 
+                if (hasOnlyLiterals && simplifiedExpressionValues.size() == valueList.getValues().size()) {
+                    // do not create a new instance of InPredicate expression if it would be same as original
+                    return node;
```

Comment:
- document what's the benefit (i guess it has something to do with instance-based caching)
- why is this specific to all-literals only? if `process()` returned equal/same Expression (for every argument), we can preserve `node` regardless of `hasOnlyLiterals`

Reply:
> why is this specific to all-literals only? if process() returned equal/same Expression (for every argument), we can preserve node regardless of hasOnlyLiterals

I though that expression comparison might be expensive, but since we do `.distinct()` above, I don't think it matters. Changed the code